pyccl/baryons.py

```diff
@@ -0,0 +1,137 @@
+from . import ccllib as lib
+from .base import CCLAutoreprObject
+from .pyutils import check
+from .pk2d import Pk2D
+from .base import unlock_instance
+import numpy as np
+
+
+class BaryonicEffects(CCLAutoreprObject):
+    """`BaryonicEffect` obects are used to include the imprint of baryons
+    on the non-linear matter power spectrum. Their main ingredient is a
+    method `apply_baryons` that takes in a `ccl.Pk2D` and returns another
+    `ccl.Pk2D` object that now accounts for baryonic effects.
+    """
+
+    def apply_baryons(self, cosmo, pk2d, in_place=False):
+        """Apply baryonic effects to a given power spectrum.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): Cosmological parameters.
+            pk2d (:class:`~pyccl.pk2d.Pk2D`): power spectrum.
+            in_place (:obj:`bool`): if True, the `pk2d` itself is modified,
+                instead of returning a new `ccl.Pk2D` object.
+        """
+        raise NotImplementedError("Do not use base `BaryonicEffects` class.")
+
+    @unlock_instance(mutate=True, argv=2)
+    def _new_pk(self, cosmo, pk2d, a_arr, lk_arr, pk_arr, logp, in_place):
+        # Returns a new Pk2D, either in place or from scratch.
+        if in_place:
+            lib.f2d_t_free(pk2d.psp)
+            status = 0
+            pk2d.psp, status = lib.set_pk2d_new_from_arrays(
+                lk_arr, a_arr, pk_arr.flatten(),
+                int(pk2d.extrap_order_lok),
+                int(pk2d.extrap_order_hik),
+                int(logp), status)
+            check(status, cosmo)
+            return None
+        else:
+            new = Pk2D(a_arr=a_arr, lk_arr=lk_arr, pk_arr=pk_arr,
+                       is_logp=logp,
+                       extrap_order_lok=pk2d.extrap_order_lok,
+                       extrap_order_hik=pk2d.extrap_order_hik)
+            return new
+
+
+class BaryonicEffectsBCM(BaryonicEffects):
```

Comment:
If we have moved `cosmo` to `__init__` and have defined `self.lk_arr` and `self.a_arr` using the sampling rates in `cosmo`, we can optionally add the following function to the class, which will return a `Pk2D` object with the correction. Users can then simply multiply the two `Pk2D` objects to get the corrected power spectrum.

And the don't have to worry about `pk.get_spline_arrays()` and array operations, because everything is already coded in for them.

```python
    def get_power_spectrum(self):
        pk = self.boost_factor(np.exp(self.lk_arr), self.a_arr)
        return Pk2D(a_arr=self.a_arr, lk_arr=self.lk_arr, pk_arr=np.log(pk),
                    is_logp=True, extrap_order_lok=0, extrap_order_hik=0)
```

So, they'd just do `pk * bar.get_power_spectrum()`

Reply:
As above, I disagree on coupling cosmo to this